sdks/python/tox.ini

```diff
@@ -160,16 +160,17 @@ commands =
 [testenv:py27-lint]
 # Checks for py2 syntax errors
 deps =
+  -r build-requirements.txt
```

Comment:
This is required if we want `tox` to work on its own, and I really really do.  Using gradle to run tox tasks adds a lot of overhead and I kind of hate it.  gradle creates a complete virtualenv just to build an sdist, then it creates another virtualenv to run tox (and IIRC it does this every time).  If I run tox directly it creates a single virtualenv for both sdist and the tox env and it reuses that virtualenv between runs.  Gradle also just seems slow to start up.  Also, I can't pass individual tests to run when using gradle.  Let's just say that as a python developer, the gradle experience is the thing I like least about the Beam project.

Yes, by adding the gen_proto requirements to `build-requirements.txt` and installing that in  the `setupVirtualenv` task I'm making that overhead worse, but I certainly didn't start that problem:  the code is currently installing `tox` every time that the `sdist` task is run, which is pointless.  In other words, `setupVirtualenv` is currently trying to be the one virtualenv used by all tasks, but different tasks have different requirements.  

I could add a feature to `setupVirtualenv` to provide a requirements file, which would let us cut out `tox` from the `sdist` task which would probably end up in a net overall win, because tox has a lot of deps.  Should I do that here or in a followup PR?

I've been meaning to bring up some of these issues on the list.  I assume the reason for the two step test process (sdist + tox, each with its own virtualenv) is to ensure that tox is using an sdist built in the same way every time, using python2, just as it would be for distribution?

Reply:
`tox` runs outside of gradle. (At least for me). What is the issue there?

> I assume the reason for the two step test process (sdist + tox, each with its own virtualenv) is to ensure that tox is using an sdist built in the same way every time, using python2, just as it would be for distribution?

sdist step is used for other things (for building artifacts for release). That is the reason for it to exists on its own.

tox, I do not believe we spent time on configuring it to reuse existing virtualenv. If that is an option we can try that.

Reply:
> tox runs outside of gradle. (At least for me). What is the issue there?

When tox is run on jenkins, it is invoked via gradle.  The tox task is defined in BeamModulePlugin.groovy like this:

```groovy
      project.ext.toxTask = { name, tox_env ->
        project.tasks.create(name) {
          dependsOn 'setupVirtualenv'
          dependsOn ':sdks:python:sdist'

          doLast {
            // Python source directory is also tox execution workspace, We want
            // to isolate them per tox suite to avoid conflict when running
            // multiple tox suites in parallel.
            project.copy { from project.pythonSdkDeps; into copiedSrcRoot }

            def copiedPyRoot = "${copiedSrcRoot}/sdks/python"
            def distTarBall = "${pythonRootDir}/build/apache-beam.tar.gz"
            project.exec {
              executable 'sh'
              args '-c', ". ${project.ext.envdir}/bin/activate && cd ${copiedPyRoot} && scripts/run_tox.sh $tox_env $distTarBall"
            }
          }
          inputs.files project.pythonSdkDeps
          outputs.files project.fileTree(dir: "${pythonRootDir}/target/.tox/${tox_env}/log/")
        }
      }
```

and it's used here:

https://github.com/apache/beam/blob/master/sdks/python/test-suites/tox/py2/build.gradle

> sdist step is used for other things (for building artifacts for release). That is the reason for it to exists on its own.

I'm not saying that `sdist` task shouldn't exist, I'm saying it adds overhead to use the `sdist` task to generate the tarball for the tox task, instead of just letting tox generate its own sdist tarball, which it is designed to do and is the default behavior (Also, there's the copying of source files which adds more overhead and it's unclear why this exists since tox is designed to create isolated virtualenvs per task environment).

In the definition of `toxTask` above, you see that it depends on `setupVirtualenv` and also on `sdist`, and `sdist` itself depends on `setupVirtualenv`.   That's why we end up running  `setupVirtualenv` twice for each tox task run, and then tox creates yet another virtualenv to actually run the task.  That's 3 virtualenvs per tox job!   I've used travisCI to setup a number of tox-based CI jobs and the typical workflow is to ensure that tox is installed *on the worker image* and then theres only one virtualenv created per job, the one created by tox. 

Anyway, long story short the `build-requirements.txt` line is required for the tox tasks to work.

Reply:
I agree that the gradle config is not the most efficient. The base requirements for running Python tests on gradle are pip, virtualenv, and 4 versions of the python interpreter. Tox is installed by setupVirtualenv. :sdks:python:sdist is run only once (IIUC) for all toxTasks, since running more than one sdist concurrently is racey and makes our tests flakey.
CC: @markflyhigh who set this up.

Thinking about this now, I have a hunch that the tarball in use doesn't ever include Cythonized sources (since they are created by sdist when the cython package is preset).

Source files are copied due to isolation requirements. The Cythonization process adds files which are automatically picked up if present. Perhaps a better way to do this is to exclude C files from sdist for the non-cython envs, but this is the way we do it currently.

Reply:
So we agree there's room for improvement.  The question for the moment is, is my current solution ok?